src/log-store/src/kafka/util/record.rs

```diff
@@ -0,0 +1,608 @@
+    }
+}
+
+fn record_type(seq: usize, num_records: usize) -> RecordType {
+    if seq == 0 {
+        RecordType::First
+    } else if seq == num_records - 1 {
+        RecordType::Last
+    } else {
+        RecordType::Middle(seq)
+    }
+}
+
+fn build_records(entry: EntryImpl, max_record_size: usize) -> Vec<Record> {
+    if entry.data.len() <= max_record_size {
+        let record = Record {
+            meta: RecordMeta {
+                version: VERSION,
+                tp: RecordType::Full,
+                entry_id: entry.id,
+                ns: entry.ns,
+                checksum: crc32fast::hash(&entry.data),
+            },
+            data: entry.data,
+        };
+        return vec![record];
+    }
+
+    let chunks = entry.data.chunks(max_record_size);
+    let num_chunks = chunks.len();
+    chunks
+        .enumerate()
+        .map(|(i, chunk)| Record {
+            meta: RecordMeta {
+                version: VERSION,
+                tp: record_type(i, num_chunks),
+                entry_id: entry.id,
+                ns: entry.ns.clone(),
+                checksum: crc32fast::hash(chunk),
+            },
+            data: chunk.to_vec(),
+        })
+        .collect()
+}
+
+pub fn maybe_emit_entry(
+    record: Record,
+    entry_records: &mut HashMap<EntryId, Vec<Record>>,
+) -> Result<Option<EntryImpl>> {
+    ensure!(
+        record.meta.checksum == crc32fast::hash(&record.data),
+        ChecksumMismatchedSnafu
+    );
+
+    let mut entry = None;
+    match record.meta.tp {
+        RecordType::Full => {
+            entry = Some(EntryImpl::from(vec![record]));
+        }
+        RecordType::First => {
+            ensure!(
+                !entry_records.contains_key(&record.meta.entry_id),
+                IllegalSequenceSnafu {
+                    error: "First record must be the first"
+                }
+            );
+            entry_records.insert(record.meta.entry_id, vec![record]);
+        }
+        RecordType::Middle(seq) => {
+            let Some(prefix) = entry_records.get_mut(&record.meta.entry_id) else {
+                return IllegalSequenceSnafu {
+                    error: "Middle record must not be the first",
+                }
+                .fail();
+            };
+            // Safety: the records are guaranteed not empty if the key exists.
+            let last_record = prefix.last().unwrap();
+            let legal = match last_record.meta.tp {
+                // Legal if this record follows a First record.
+                RecordType::First => seq == 1,
+                // Legal if this record follows a Middle record just prior to this record.
+                RecordType::Middle(last_seq) => last_seq + 1 == seq,
+                // Illegal sequence.
+                _ => false,
+            };
+            ensure!(
+                legal,
+                IllegalSequenceSnafu {
+                    error: "Illegal prefix for a Middle record"
```

Comment:
Can we provide more info here for debugging?

Reply:
I think people does not need to how the sequence is regarded illegal. Just saying it's illegal would be enough.